### ISSUE TITLE
Better HDR merge with auto-allignment

### ISSUE THREAD

**Issue by gluckgluckwasserbauch** [darktable-org/darktable#17326]
**Is your feature request related to a problem? Please describe.**

I like to shoot HDR images. For that, I use exposure bracketing and shoot 3 images at once. 
One underexposed, one optimally, and one overexposed. 
Then I stack them onto each other to create an HDR image.

I didn't use Adobe Lightroom yet, but from what I've heard, it does that way better than Darktable.
The main reason: it auto-alligns the images to fit perfectly onto each other.
With DT, if I don't shoot with a tripod, they don't get stacked onto each other, which causes a very janky picture, even with super fast shutter speeds and a steady hand (see below).

There are a few LUA scripts that do that, but I couldn't get them to work.
(And, tbh, having to resort to janky scripts for a basic feature like that shouldn't be standard imo).

Also, the colors are quite a bit off and the whole image looks very artificial, even if it would be stacked correctly.

**Describe the solution you'd like**

My idea would be to add a dialog window after clicking the "create HDR picture" button.
In the dialog, you can generate a preview and have some toggles, like 'Auto-allign if possible', 'Set transparency' for every frame, and maybe some more.  
It should also be possible to safe those preferences and apply them automatically.

This feature could be relatively easy to implement. It is basically just a GUI-frontend for some already existing scripts, like Hugin, ImageMagick, Enfuse, and maybe some more.

**Alternatives**
- I tried a few scripts before, but Hugin (and a few others) failed, and those that worked, didn't create a usable picture.
- There's also [Luminance HDR](https://flathub.org/apps/net.sourceforge.qtpfsgui.LuminanceHDR), but that threw off the colors the one time I tried it.
- I used [Conversion Tool](https://www.conversion-tool.com/hdrmerger) for the sample photo, but even that didn't look great.

**Additional context**

*How it should look like:* (kind of)
![DSC00549](https://github.com/user-attachments/assets/ae54a8a0-fd04-4266-a19f-16df49546274)
*How it actually looks like in DT:* 
![Bildschirmfoto_20240814_183711](https://github.com/user-attachments/assets/8157205e-e5ee-49ff-aaf1-1bb6e2697d35)

**Comment by wpferguson** [comment]
> (And, tbh, having to resort to janky scripts for a basic feature like that shouldn't be standard imo).

darktable is a raw processor, not an HDR tool  HDRmerge, enfuse, etc. are HDR tools.  Lua scripting is built into darktable to allow extended functionality for things that darktable doesn't, can't, wont, provide natively.

> There are a few LUA scripts that do that, but I couldn't get them to work.

Did you try raising an issue or asking on forums?

**Comment by jenshannoschwalm** [comment]
@wpferguson we might want to discuss "remove the dt HDR merge code" as of too-low quality?

**Comment by wpferguson** [comment]
IIRC I tried the HDR in darktable 3 cameras ago (~8 years).  I created the HDR from 3 images, but then you had to apply the global tonemap module to get something usable.  I never had much luck using it.  Houz had written the official/enfuse.lua script by then and I used that and got good results.

Since global tonemap is deprecated I'm not sure there is a way to get a usable HDR from the darktable HDR merge.

**Comment by wpferguson** [comment]
![hdr](https://github.com/user-attachments/assets/2413aaf4-4260-4ced-b5f1-a74fce08d04a)

The left is darktable HDR with scene referred defaults.  The middle is HDRmerge.  The right is enfuse.  I didn't edit any of them, so this is just the output of the process

**Comment by wpferguson** [comment]
This is enfuseAdvanced output

![efa](https://github.com/user-attachments/assets/9653fcf5-51af-440a-9a5a-446d7f3262cb)

**Comment by wpferguson** [comment]
HDRmerge and enfuseAdvanced align the images.  enfuse doesn't, but it could be added to the script with a couple of lines of code.  darktable does not align the images.

**Comment by wpferguson** [comment]
HDRMerge can take raw files, so no export required and produces a DNG file.  So essentially it is the same as darktable HDR but with image auto align.  

So this means that we could "replace" the `Create HDR` button with another `Create HDR` button other than the button would be the full width of the module instead of half width and run a script that runs HDRMerge.

**Comment by wpferguson** [comment]
I played :smile:

Here's a handheld HDR bracket merged with the current darktable Create HDR

![hdr](https://github.com/user-attachments/assets/83083564-65c5-4804-9a52-14377799c524)


And here's the same bracket with HDRMerge triggered from a script

![hdrmerge](https://github.com/user-attachments/assets/706f6543-0254-4fba-beae-ff5c8af22bc1)

You can see the misalignment in his head and especially in the fence, which is fixed with HDRMerge.

The script works, but it needs a couple of changes to provide more feedback to the user.

So what do we need to do to make this a replacement for the current Create HDR?

**We need HDRMerge**

Do we....

- Tell the user that darktable now uses HDRMerge to create HDRs and it's the user's responsibility to install it?  Packages are available for Linux, Windows, and MacOS so it shouldn't be a problem, but it is a step.
- Include HDRMerge as a binary in the darktable bin dir?
- Add it as a subproject and build it along with darktable?

**The lua-scripts need to be part of the darktable package**

- If we are going to rely on the scripts to provide "core" functionality then they  should be distributed as part of the darktable package instead of relying on the user to install them.  In 4.7 @TurboGit added the lua-scripts as a subproject so that we could work on using darktable's translation facilities for the scripts.  We didn't finish it in time so we pulled it back.  We're ready to start on it again.  Yesterday I played with packaging the scripts as part of darktable and got it working.  It's still a work in progress, but it can be done.
- Scripts that provide "core" functionality should be put in a "system" directory and started automatically, i.e. they should "just work".  I'm also working on a script that persists grouping across database instances and it would be a candidate for this.

**Comment by JLTastet** [comment]
I would like to second @gluckgluckwasserbauch’s enhancement request.

I also use HDR merge a lot, but with a slightly different goal: I have started editing some of my photos for high dynamic range displays (think Rec. 2100), exporting them to HDR formats like JPEG XL, AVIF, UltraHDR JPEG, etc.

For HDR edits, it is critical to preserve the highlights, while still having decently clean shadows. But, sometimes, even the dynamic range of a modern full-frame camera isn’t sufficient for this, and I will use the burst mode to perform some basic handheld bracketing (often just an extra -3EV exposure) to recover some dynamic range. [This pixls.us post](https://discuss.pixls.us/t/manual-creation-of-ultrahdr-images/45004/20?u=jltastet) contains an image produced this way.

Having the HDR merge feature directly accessible in Darktable is very convenient, since it produces in one click a DNG file that can be edited just like any other raw file, but with increased dynamic range.

However, while I sometimes obtain good results if my hands remain steady and the IBIS does a good job (or I brought my tripod), most of the time the shots will be shifted/rotated by a few pixels, leading to a low "keepers" rate.

I can usually recover the shots by aligning them with Hugin while preserving the full dynamic range. To do so, I use EXR as input and output format, and I disable any exposure blending/compensation and tone mapping. This also requires applying all corrective modules before merging, and manually managing the colour spaces (I do everything in linear Rec. 2020), since EXR doesn’t store this data. So, while certainly doable, the process is quite cumbersome. I have also tried the Lua script, but this didn’t work at all for me, producing completely wrong colours.

All that to say that having a working HDR merge button with auto-alignment within Darktable could be very useful, regardless of whether it is implemented in Darktable itself or outsourced to a Lua plugin (just not the current one, which seems broken). For me, the most important part is to have the option to disable any tone mapping and get a scene-referred, (perceptually) lossless, "raw-like" output, like a demosaiced DNG or another high-bit-depth file.

Just my two cents.

**Comment by da-phil** [comment]
> I would like to second [@gluckgluckwasserbauch](https://github.com/gluckgluckwasserbauch)’s enhancement request.
> 
> I also use HDR merge a lot, but with a slightly different goal: I have started editing some of my photos for high dynamic range displays (think Rec. 2100), exporting them to HDR formats like JPEG XL, AVIF, UltraHDR JPEG, etc.
> 
> For HDR edits, it is critical to preserve the highlights, while still having decently clean shadows. But, sometimes, even the dynamic range of a modern full-frame camera isn’t sufficient for this, and I will use the burst mode to perform some basic handheld bracketing (often just an extra -3EV exposure) to recover some dynamic range. [This pixls.us post](https://discuss.pixls.us/t/manual-creation-of-ultrahdr-images/45004/20?u=jltastet) contains an image produced this way.
> 
> Having the HDR merge feature directly accessible in Darktable is very convenient, since it produces in one click a DNG file that can be edited just like any other raw file, but with increased dynamic range.
> 
> However, while I sometimes obtain good results if my hands remain steady and the IBIS does a good job (or I brought my tripod), most of the time the shots will be shifted/rotated by a few pixels, leading to a low "keepers" rate.
> 
> I can usually recover the shots by aligning them with Hugin while preserving the full dynamic range. To do so, I use EXR as input and output format, and I disable any exposure blending/compensation and tone mapping. This also requires applying all corrective modules before merging, and manually managing the colour spaces (I do everything in linear Rec. 2020), since EXR doesn’t store this data. So, while certainly doable, the process is quite cumbersome. I have also tried the Lua script, but this didn’t work at all for me, producing completely wrong colours.
> 
> All that to say that having a working HDR merge button with auto-alignment within Darktable could be very useful, regardless of whether it is implemented in Darktable itself or outsourced to a Lua plugin (just not the current one, which seems broken). For me, the most important part is to have the option to disable any tone mapping and get a scene-referred, (perceptually) lossless, "raw-like" output, like a demosaiced DNG or another high-bit-depth file.
> 
> Just my two cents.
> 
> EDIT: I now remember the issue with HDRmerge. The stable release is _massively_ outdated (2015) and produces incorrect colors, while newer development versions are only available for Linux (but not Windows or macOS). So using it instead of the current "Create HDR" would amount to dropping support for this feature on two platforms.

I agree, I've been using HDRmerge as well, to overcome the limitations of the HDR merge process in darktable.
But it's not a great workflow.

I wanted to look into the topic of auto-alignment for HDR merge myself, and see how far I can get.
Probably I will start with a simple approach, only taking 2D alignment (shift in x & y + roll rotation) into account. This should cover most exposure brackets, at least for me, as most exposure brackets are either shot on a (shaky) tripod or with fast shutter speeds out of my hand. Then most of the tiny 6D pose differences between all shots can be estimated with the above mentioned 2D alignments. And then let's see if we need more than that...

**Comment by da-phil** [comment]
This issue is still relevant.

**Comment by da-phil** [comment]
FYI, I'm already working on this feature, however it's still in prototype state and does not work well on all sort of alignment challenges.
Once I have something remotely usable for testing, I'll ping you guys again.

**Comment by wpferguson** [comment]
@da-phil why not look at the HDRmerge source and see how they did image alignment, or look at align_image_stack from hugin?

**Comment by da-phil** [comment]
> [@da-phil](https://github.com/da-phil) why not look at the HDRmerge source and see how they did image alignment, or look at align_image_stack from hugin?

There is a couple of ways we can do it, and the way HDRmerge aligns images is quite primitive, it only consideres translations in x & y, not rotations and perspective distortions (part of a homography) which are required for alignment of free-hand image stacks.

I consider align_image_stack unusable for the purposes of aligning RAW images for a RAW HDR merge function as it can neither read, nor write RAW images and even if it could, using an external tool with an own RAW conversion functionality is not a good idea. The novelty of the core algorithm of darktable's HDR merge is that it merges RAW files within the RAW domain, without the necessity to demosaic the images first.

In the meantime I've been working on a python program, which utilizes OpenCV to align even images which are taken free-hand for various purposes (HDR & focus stacking and astro sky image alignment). I wanted to take this prototype as a starting point for a C implementation.

**Comment by wpferguson** [comment]
> which are required for alignment of free-hand image stacks.

I used HDRmerge to successfully merge my hand held bracket shots with no problem.  I think for simple 3 or 5 image brackets HDRmerges x y approach works just fine.  If we want to get into focus stacking or astro alignment then that's another issue that should be addressed separately.

**Comment by da-phil** [comment]
> If we want to get into focus stacking or astro alignment then that's another issue that should be addressed separately.

Yes, certainly, this is definitely out-of-scope for this feature. However, just correcting simple translations has only worked for my bracketed shots which were taken on a (shaky) tripod, but not so well for my handheld shots which mostly needed rotational or even perspective corrections to achieve a good alignment.

My general RFC question would be: are we willing to accept another open-source lib ([OpenCV](https://opencv.org/)) as dependency for darktable? The working python implementation could be translated to C in a rather straight forward fashion when we stick to using OpenCV , which is used for the following:
* detecting SIFT features
* description of SIFT features
* descriptor matching
* robust transform estimation using RANSAC

There is probably also other ways with other - smaller - libraries to achieve that, but not necessarily simpler. Not using any library for the above mentioned steps would mean re-writing and maintaining a very large portion of code in the ten-thousands lines of code range.

**Comment by jenshannoschwalm** [comment]
`OpenCV` is **very** large.

**Comment by jenshannoschwalm** [comment]
BTW - rawprepare has a nice aligning algo ...

**Comment by da-phil** [comment]
> `OpenCV` is **very** large.

yes, indeed, that's why I was asking how people feel about it.


> BTW - rawprepare has a nice aligning algo ...

Oh nice, I did not know, can you point me to the actual code?

To be honest, the implementation of a good and robust image alignment algorithm which also works for handheld image stacks under extreme exposure differences is probably going to be one of the most complex algorithms in the entire DT repo. The sheer complexity of such an algorithm is the reason why most open source image editing tools are lacking this functionality, also GIMP still does not have this functionality, whereas in commercial tools it's just a click on your image stack processing menu.

**Comment by jenshannoschwalm** [comment]
Of course i meant rawtherapee :-) They had aligning code for demosaicing multi-image files ...

**Comment by da-phil** [comment]
Aligning multi-image files for demosaicing is again a different use-case than aligning images taken on a shaky tripod or even handsfree. That code probably even uses assumptions about the applied sensor shifts in x and y, however will not consider other transformation components (rotations, perspective distortions) which are unreasonable for deterministic sensor shift functionality for increasing resolution and/or increasing signal-to-noise ratio or color fidelity of multi-image RAW files.

In the meantime I'll let claude do a cheap proof-of-concept to see how complex a C implementation using OpenCV is going to be and how well it works for folks who requested the functionality. And I keep an eye out for alternatives which could avoid this heavy dependency...

**Comment by wpferguson** [comment]
> aligning images taken on a shaky tripod or even handsfree

I think x y alignment is probably sufficient for an HDR bracket set.  I've shot a lot of those hands free and never had an issue aligning them with HDRmerge or align_image_stack.  I especially can't image a tripod set requiring more transformation that that.  

How much effort and code do we want to spend trying to rescue something, which if it requires that amount of transformation, probably isn't worth rescuing anyway?

**Comment by da-phil** [comment]
> I've shot a lot of those hands free and never had an issue aligning them with HDRmerge or align_image_stack.

Maybe your quality requirements are different from mine, but I was never happy with HDRmerge results in such situations, and I know that you can do better, using align_image_stack and my own prototype, which are able to perfectly align most images (if there are no parallax issues).

> How much effort and code do we want to spend trying to rescue something, which if it requires that amount of transformation, probably isn't worth rescuing anyway?

I'm not sure if "rescue" is the right terminology here, given that "computational photography" has made this workflow an actual regular use-case. At the same time I'm not saying we should go full-in on "computational photography", but at least offer a similar amount of functionality such as commercial software (e.g. photoshop & lightroom). I can only tell you that only aligning displacements in x and y do not work for 99% of my free hand shots, the misalignment is very visible when stacking images in gimp using the "difference" layer function, or just doing a HDR merge and having nasty ghosting in some areas, particularly in corner areas.

**Comment by wpferguson** [comment]
How do you shoot a set of images for HDR?  I've always used the in camera settings to set up a 3 or 5 image bracket and then just triggered off the 3 or 5 images in a second or so.  I guess if you had to do some manual adjustment that would cause bigger alignment problems.

**Comment by jenshannoschwalm** [comment]
I had a look again at rawtherapee code, found it in `pixelshift.cc`, there is also some introduction in  https://pixls.us/articles/rawtherapee-and-pentax-pixel-shift/ and plenty more around.

**Comment by da-phil** [comment]
> How do you shoot a set of images for HDR? I've always used the in camera settings to set up a 3 or 5 image bracket and then just triggered off the 3 or 5 images in a second or so. I guess if you had to do some manual adjustment that would cause bigger alignment problems.

I set my camera (Sony A7r5) to HDR bracket mode with image burst mode which rarely takes more than a second during the day. Maybe the difference is also in the resolution, the 61 MP of my camera are unforgiving in available light photography and even on the a shaky compact tripod, when doing 1/10 - 5s exposures out in the wind. I didn't see most of those issues when I still had a 20 MP camera 10 years ago. You can simply calculate how much pixel shift you get when you have only a very small rotation component in the deviation between frames, in the corners this will quickly add up to a two digit pixel misalignment and this is not going to be fixed with only a transformation in x and y.

**Comment by da-phil** [comment]
I created a new draft PR which includes an auto-alignment implementation using OpenCV: https://github.com/darktable-org/darktable/pull/21420
I've been testing this version for quite some time on many "shaky tripod" and handheld exposure stacks 
with very good results. 

I'll run the Nightly PKG action and create app images for folks to try out, stay tuned ;)

**Comment by jandren** [comment]
This is awesome!
I definently think full fit needs to be done to make something like this actually good. My usecase would be for low light situations where image stacking is more for giving than long exposure.

I personally have a dream of doing this with custom tracking, for example lock onto the eyes of a subject and position/rotate around them (or just target the whole face) could produce some nice results where the subject is in motion. Might be out of scope for this attempt but technically quite similar.

**Comment by da-phil** [comment]
> I personally have a dream of doing this with custom tracking, for example lock onto the eyes of a subject and position/rotate around them (or just target the whole face) could produce some nice results where the subject is in motion. Might be out of scope for this attempt but technically quite similar.

I'm not sure if I understood this correctly, but you meant really a face tracking algorithm for a camera on a gimbal or something like this? Or is this also related to merging images into a HDR?

**Comment by jenshannoschwalm** [comment]
@da-phil got some really good results!

**Comment by da-phil** [comment]
@gluckgluckwasserbauch @JLTastet 
I'm still waiting for feedback from you, as you requested / co-requested this feature.
The work has been done in this PR: https://github.com/darktable-org/darktable/pull/21420
And here are pre-built OS packages for testing: https://github.com/da-phil/darktable/actions/runs/28748026727#artifacts
